src/core/entity.js

```diff
@@ -55,44 +59,48 @@ class Entity {
     }
   }
 
-  look_at(vec, up = this.up) {
-    math.mat4.target_to(this.local_matrix, this.position, vec, up);
-  }
-
-  get left() {
-    const out = this.local_matrix.slice(0, 3);
-    return math.vec3.normalize(out, out);
-  }
-
   get up() {
-    const out = this.local_matrix.slice(4, 7);
+    const out = this.matrix.slice(4, 7);
     return math.vec3.normalize(out, out);
   }
 
   get forward() {
-    const out = this.local_matrix.slice(8, 11);
+    const out = this.matrix.slice(8, 11);
     return math.vec3.normalize(out, out);
   }
 
   set position(vec) {
-    const offset = [];
-    math.vec3.subtract(offset, vec, this.position);
-    math.mat4.translate(this.local_matrix, this.local_matrix, offset);
+    math.mat4.compose(
+      this.matrix, this.rotation, vec, this.scale, this.origin
+    );
   }
 
   get position() {
-    return this.local_matrix.slice(12, 15);
+    return this.matrix.slice(12, 15);
   }
 
   set scale(vec) {
     this._scale = vec;
-    math.mat4.scale(this.local_matrix, this.local_matrix, vec);
+    math.mat4.compose(
+      this.matrix, this.rotation, this.position, vec, this.origin
+    );
   }
 
   get scale() {
     return this._scale;
   }
 
+  set rotation(quaternion) {
+    this._rotation = quaternion;
+    math.mat4.compose(
+      this.matrix, quaternion, this.position, this.scale, this.origin
+    );
+  }
+
+  get rotation() {
+    return this._rotation;
```

Comment:
I'm not a big fan of these getters. We need them because we need setters. What if we had a separate `update_matrix` or `apply` method which has to be explicitly called by the user every time `scale`, `rotation` or `position` change?

Reply:
I'm not a huge fan of those. Maybe we can do `jQuery` style getters/setter? `Entity.scale([1, 1, 1]);` sets a value, and `Entity.scale()` returns it? With this approach private attributes like `Entity._rotation` can be stored in the method's closure and won't be exposed anywhere.

Reply:
That won't work with Entity being a class, right?

Reply:
Unless you define those methods in the `constructor`.

Reply:
You're roite Sir... Let's leave this for now.